// @looker/filter-components/Filter/components/AdvancedFilter/components/MidInputLabel/MidInputLabel.js
import { inputHeight, Text } from '@looker/components';
import styled from 'styled-components';
export var MidInputLabel = styled(Text).attrs(function (_ref) {
  var _ref$color = _ref.color,
      color = _ref$color === void 0 ? 'text1' : _ref$color,
      _ref$fontSize = _ref.fontSize,
      fontSize = _ref$fontSize === void 0 ? 'small' : _ref$fontSize,
      _ref$px = _ref.px,
      px = _ref$px === void 0 ? 'xsmall' : _ref$px,
      _ref$textAlign = _ref.textAlign,
      textAlign = _ref$textAlign === void 0 ? 'right' : _ref$textAlign;
  return {
    color: color,
    fontSize: fontSize,
    px: px,
    textAlign: textAlign
  };
}).withConfig({
  displayName: "MidInputLabel",
  componentId: "sc-krsmua-0"
})(["align-items:center;align-self:center;background:", ";border:1px solid ", ";border-right:none;display:flex;height:", ";line-height:", ";text-overflow:ellipsis;white-space:nowrap;"], function (_ref2) {
  var colors = _ref2.theme.colors;
  return colors.background;
}, function (_ref3) {
  var colors = _ref3.theme.colors;
  return colors.ui3;
}, inputHeight, function (_ref4) {
  var theme = _ref4.theme;
  return theme.lineHeights.medium;
});
