// @looker/visualizations-adapters/__mocks__/mockQueryContextValues.js
import { mockBarConfig } from './mockConfig';
import { mockFields } from './mockFields';
export var mockQueryContextValues = {
  config: mockBarConfig,
  data: [{
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 1088,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2515,
    'orders.average_total_amount_of_order_usd': 1069,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }],
  fields: mockFields,
  ok: true,
  loading: false
};
export var mockContextWithNull = {
  config: mockBarConfig,
  data: [{
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 1088,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2087,
    'orders.average_total_amount_of_order_usd': 88,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': null,
    'orders.average_total_amount_of_order_usd': null,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 88,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2515,
    'orders.average_total_amount_of_order_usd': 1069,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }],
  fields: mockFields,
  ok: true,
  loading: false
};
