// @looker/visualizations-visx/Axis/XAxis.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import "core-js/modules/es.array.index-of.js";
var _excluded = ["formattedValue"];
import "core-js/modules/es.object.define-property.js";

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { useContext } from 'react';
import { Text } from '@visx/text';
import { Axis, ThemeContext as VisxThemeContext, DataContext } from '@visx/xychart';
import { formatDateLabel } from '../utils/formatDateLabel';
export var XAxis = function XAxis(_ref) {
  var fields = _ref.fields,
      label = _ref.label,
      labelDy = _ref.labelDy,
      showTicks = _ref.showTicks,
      tickAngle = _ref.tickAngle,
      tickTextAnchor = _ref.tickTextAnchor,
      tickSpace = _ref.tickSpace;
  var visxTheme = useContext(VisxThemeContext);

  var _useContext = useContext(DataContext),
      _useContext$width = _useContext.width,
      width = _useContext$width === void 0 ? 0 : _useContext$width;

  var numTicks = Math.floor(width / tickSpace);
  return React.createElement(Axis, {
    hideTicks: !showTicks,
    label: label,
    labelOffset: showTicks ? undefined : 0,
    labelProps: _objectSpread(_objectSpread({}, visxTheme.axisStyles.x.bottom.axisLabel), {}, {
      dy: labelDy
    }),
    orientation: "bottom",
    numTicks: numTicks,
    tickComponent: function tickComponent(_ref2) {
      var formattedValue = _ref2.formattedValue,
          tickProps = _objectWithoutProperties(_ref2, _excluded);

      return showTicks ? React.createElement(Text, tickProps, formatDateLabel({
        dateString: formattedValue || '',
        fields: fields
      })) : null;
    },
    tickLabelProps: function tickLabelProps() {
      return {
        angle: tickAngle,
        textAnchor: tickTextAnchor
      };
    }
  });
};
