// @looker/visualizations-adapters/VisWrapper/VisWrapper.js
import "core-js/modules/es.object.assign.js";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useContext, forwardRef } from 'react';
import styled, { css, ThemeContext } from 'styled-components';
import { ComponentsProvider } from '@looker/components';
var VisWrapperInternal = forwardRef(function (props, ref) {
  var theme = useContext(ThemeContext);

  if (!theme) {
    return React.createElement(ComponentsProvider, null, React.createElement(VisWrapperInternal, _extends({}, props, {
      ref: ref
    })));
  }

  return React.createElement("div", _extends({}, props, {
    ref: ref
  }));
});
VisWrapperInternal.displayName = 'VisWrapperInternal';

var flexDirection = function flexDirection(_ref) {
  var legend = _ref.legend;
  var positionMap = {
    top: 'column-reverse',
    right: 'row',
    left: 'row-reverse',
    bottom: 'column'
  };
  var position = legend ? legend.position : 'bottom';
  return css(["flex-direction:", ";"], positionMap[position]);
};

export var VisWrapper = styled(VisWrapperInternal).withConfig({
  displayName: "VisWrapper",
  componentId: "sc-sssqk1-0"
})(["display:flex;flex:1;", " & > div > svg{overflow:visible;}"], flexDirection);
