// @looker/filter-components/cjs/Filter/components/AdvancedFilter/components/NumberFilter/components/Between/Between.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Between = void 0;

var _components = require("@looker/components");

var _react = _interopRequireDefault(require("react"));

var _reactI18next = require("react-i18next");

var _GroupInput = require("../../../GroupInput");

var _GroupSelect = require("../../../GroupSelect");

var _MidInputLabel = require("../../../MidInputLabel");

var _utils = require("../../../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Between = ({
  item,
  onChange
}) => {
  const _useTranslation = (0, _reactI18next.useTranslation)('Between'),
        t = _useTranslation.t;

  const betweenOptions = (0, _utils.useBetweenOptions)();
  const _item$high = item.high,
        high = _item$high === void 0 ? '' : _item$high,
        _item$low = item.low,
        low = _item$low === void 0 ? '' : _item$low,
        id = item.id;

  const lowChange = event => {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      low: event.target.value
    });
  };

  const highChange = event => {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      high: event.target.value
    });
  };

  const selectChange = value => {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      bounds: value
    });
  };

  return _react.default.createElement(_components.Flex, {
    flexDirection: "row"
  }, _react.default.createElement(_GroupSelect.GroupSelect, {
    placement: "middle",
    value: item.bounds,
    options: betweenOptions,
    onChange: selectChange
  }), _react.default.createElement(_components.Flex, {
    alignItems: "center"
  }, _react.default.createElement(_GroupInput.GroupInput, {
    value: low,
    type: "number",
    onChange: lowChange,
    minWidth: "4.5em",
    "data-testid": "low"
  }), _react.default.createElement(_MidInputLabel.MidInputLabel, null, t('AND')), _react.default.createElement(_GroupInput.GroupInput, {
    value: high,
    type: "number",
    onChange: highChange,
    placement: "right",
    minWidth: "4.5em",
    "data-testid": "high"
  })));
};

exports.Between = Between;
