// @looker/visualizations-adapters/cjs/Debug/Debug.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Debug = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@looker/components");

var _StaticTable = require("../StaticTable");

var _isEmpty = _interopRequireDefault(require("lodash/isEmpty"));

var _KeyValueList = require("../KeyValueList");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Debug = ({
  ok,
  data,
  error,
  config = {},
  fields
}) => {
  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_KeyValueList.KeyValueList, {
    value: {
      response: ok ? 'ok' : 'error'
    }
  }), !(0, _isEmpty.default)(error) && _react.default.createElement(_components.Accordion2, {
    indicatorPosition: "left",
    defaultOpen: true,
    label: "Error"
  }, _react.default.createElement(_KeyValueList.KeyValueList, {
    value: error || {}
  })), !(0, _isEmpty.default)(config) && _react.default.createElement(_components.Accordion2, {
    indicatorPosition: "left",
    label: "Config"
  }, _react.default.createElement(_KeyValueList.KeyValueList, {
    value: config
  })), !(0, _isEmpty.default)(fields) && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_components.Accordion2, {
    indicatorPosition: "left",
    label: "Dimensions"
  }, _react.default.createElement(_KeyValueList.KeyValueList, {
    value: (fields === null || fields === void 0 ? void 0 : fields.dimensions) || {}
  })), _react.default.createElement(_components.Accordion2, {
    indicatorPosition: "left",
    label: "Measures"
  }, _react.default.createElement(_KeyValueList.KeyValueList, {
    value: (fields === null || fields === void 0 ? void 0 : fields.measures) || {}
  }))), data && _react.default.createElement(_components.Accordion2, {
    indicatorPosition: "left",
    defaultOpen: true,
    label: "Result"
  }, _react.default.createElement(_StaticTable.StaticTable, {
    data: data,
    fields: fields
  })));
};

exports.Debug = Debug;
