// @looker/visualizations-visx/Bar/Bar.js
var _excluded = ["formattedValue"],
    _excluded2 = ["formattedValue"];

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import "core-js/modules/es.array.map.js";
import "core-js/modules/es.function.name.js";
import "core-js/modules/es.array.iterator.js";
import "core-js/modules/es.object.to-string.js";
import "core-js/modules/web.dom-collections.iterator.js";
import "core-js/modules/es.object.define-property.js";
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import "core-js/modules/es.array.index-of.js";
import React, { useContext } from 'react';
import { Axis, DataProvider, BarSeries, XYChart, BarStack, BarGroup, ThemeContext as VisxThemeContext } from '@visx/xychart';
import { Text } from '@visx/text';
import { DEFAULT_HEIGHT, VisWrapper, pickLongestLabel, useMeasuredText } from '@looker/visualizations-adapters';
import { XYLegend } from '../XYLegend';
import isArray from 'lodash/isArray';
import get from 'lodash/get';
import compact from 'lodash/compact';
import { concatDimensions, getX, getY, getYAxisRange, useChartTheme, isValidChartData, formatDateLabel } from '../utils';
import { XYTooltip } from '../XYTooltip';
import { Grid } from '../Grid';
export var Bar = function Bar(_ref) {
  var _config$y_axis, _config$x_axis;

  var data = _ref.data,
      config = _ref.config,
      _ref$height = _ref.height,
      height = _ref$height === void 0 ? DEFAULT_HEIGHT : _ref$height,
      width = _ref.width,
      fields = _ref.fields;
  var positioning = config.positioning,
      seriesList = config.series,
      legend = config.legend;
  var formattedData = concatDimensions(data, fields);
  var chartTheme = useChartTheme(seriesList);
  var visxTheme = useContext(VisxThemeContext);
  var yAxisLabels = formattedData.map(function (datum) {
    return formatDateLabel({
      dateString: datum.dimension,
      fields: fields
    });
  });
  var yAxisLongestLabel = pickLongestLabel(yAxisLabels);

  var _useMeasuredText = useMeasuredText(yAxisLongestLabel, {
    fontFamily: visxTheme.axisStyles.x.bottom.tickLabel.fontFamily || 'Roboto',
    fontSize: visxTheme.axisStyles.x.bottom.tickLabel.fontSize || '1rem'
  }),
      longestLabelWidth = _useMeasuredText.width;

  if (!isValidChartData(data, fields)) {
    return null;
  }

  var renderedBars = compact(fields.measures.map(function (measure, i) {
    var series = isArray(seriesList) ? get(config, ['series', i]) : get(config, ['series', measure.name]);
    if (!series.visible) return undefined;
    return React.createElement(BarSeries, {
      key: i,
      dataKey: measure.name,
      data: formattedData,
      xAccessor: function xAccessor(d) {
        return getY(d, i);
      },
      yAccessor: function yAccessor(d) {
        return getX(d);
      }
    });
  }));
  var domain = positioning === 'percent' ? [0, 1] : getYAxisRange({
    config: config,
    data: formattedData,
    fields: fields
  });

  var xScale = _objectSpread({
    type: 'linear'
  }, domain && {
    domain: domain,
    zero: false
  });

  var xAxisConfig = config === null || config === void 0 ? void 0 : (_config$y_axis = config.y_axis) === null || _config$y_axis === void 0 ? void 0 : _config$y_axis[0];
  var xAxis = React.createElement(Axis, {
    hideTicks: !(xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values),
    label: (xAxisConfig === null || xAxisConfig === void 0 ? void 0 : xAxisConfig.label) || '',
    labelOffset: xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values ? undefined : 0,
    labelProps: _objectSpread({}, visxTheme.axisStyles.x.bottom.axisLabel),
    orientation: "bottom",
    tickComponent: function tickComponent(_ref2) {
      var formattedValue = _ref2.formattedValue,
          tickProps = _objectWithoutProperties(_ref2, _excluded);

      return xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values ? React.createElement(Text, tickProps, formattedValue) : null;
    }
  });
  var yAxisConfig = config === null || config === void 0 ? void 0 : (_config$x_axis = config.x_axis) === null || _config$x_axis === void 0 ? void 0 : _config$x_axis[0];
  var yAxis = React.createElement(Axis, {
    hideTicks: !(yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values),
    label: (yAxisConfig === null || yAxisConfig === void 0 ? void 0 : yAxisConfig.label) || '',
    labelOffset: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? undefined : 0,
    labelProps: _objectSpread(_objectSpread({}, visxTheme.axisStyles.y.left.axisLabel), {}, {
      dx: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? -longestLabelWidth - 10 : -10
    }),
    orientation: "left",
    tickComponent: function tickComponent(_ref3) {
      var formattedValue = _ref3.formattedValue,
          tickProps = _objectWithoutProperties(_ref3, _excluded2);

      return yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? React.createElement(Text, tickProps, formatDateLabel({
        dateString: formattedValue || '',
        fields: fields
      })) : null;
    }
  });
  var chartMargin = {
    right: 0,
    top: 0,
    bottom: 50,
    left: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? longestLabelWidth + 50 : 50
  };
  return React.createElement(DataProvider, {
    xScale: xScale,
    yScale: {
      type: 'band'
    },
    theme: chartTheme
  }, React.createElement(VisWrapper, {
    legend: legend
  }, React.createElement(XYChart, {
    margin: chartMargin,
    width: width,
    height: height
  }, xAxis, yAxis, React.createElement(Grid, {
    config: config
  }), React.createElement(XYTooltip, {
    config: config,
    data: formattedData,
    fields: fields,
    showDatumGlyph: false,
    snapToDatum: false
  }), positioning === 'stacked' || positioning === 'percent' ? React.createElement(BarStack, {
    offset: positioning === 'percent' ? 'expand' : 'none'
  }, renderedBars) : React.createElement(BarGroup, null, renderedBars)), React.createElement(XYLegend, {
    config: config,
    fields: fields
  })));
};
