// @looker/filter-components/Filter/components/AdvancedFilter/components/NumberFilter/components/Between/Between.js
import { Flex } from '@looker/components';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { GroupInput } from '../../../GroupInput';
import { GroupSelect } from '../../../GroupSelect';
import { MidInputLabel } from '../../../MidInputLabel';
import { useBetweenOptions } from '../../../../utils';
export var Between = function Between(_ref) {
  var item = _ref.item,
      onChange = _ref.onChange;

  var _useTranslation = useTranslation('Between'),
      t = _useTranslation.t;

  var betweenOptions = useBetweenOptions();
  var _item$high = item.high,
      high = _item$high === void 0 ? '' : _item$high,
      _item$low = item.low,
      low = _item$low === void 0 ? '' : _item$low,
      id = item.id;

  var lowChange = function lowChange(event) {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      low: event.target.value
    });
  };

  var highChange = function highChange(event) {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      high: event.target.value
    });
  };

  var selectChange = function selectChange(value) {
    onChange === null || onChange === void 0 ? void 0 : onChange(id, {
      bounds: value
    });
  };

  return React.createElement(Flex, {
    flexDirection: "row"
  }, React.createElement(GroupSelect, {
    placement: "middle",
    value: item.bounds,
    options: betweenOptions,
    onChange: selectChange
  }), React.createElement(Flex, {
    alignItems: "center"
  }, React.createElement(GroupInput, {
    value: low,
    type: "number",
    onChange: lowChange,
    minWidth: "4.5em",
    "data-testid": "low"
  }), React.createElement(MidInputLabel, null, t('AND')), React.createElement(GroupInput, {
    value: high,
    type: "number",
    onChange: highChange,
    placement: "right",
    minWidth: "4.5em",
    "data-testid": "high"
  })));
};
