// @looker/filter-expressions/utils/location/describe_location.js
import "core-js/modules/es.array.concat.js";
import "core-js/modules/es.number.to-fixed.js";
import defaultTo from 'lodash/defaultTo';
import { describeUserAttribute } from '../user_attribute/describe_user_attribute';

var describeLat = function describeLat(lat) {
  return "".concat(Math.abs(lat).toFixed(1), "\xB0").concat(lat > 0 ? 'N' : 'S');
};

var describeLon = function describeLon(lon) {
  return "".concat(Math.abs(lon).toFixed(1), "\xB0").concat(lon > 0 ? 'E' : 'W');
};

var box = function box(_ref) {
  var lon = _ref.lon,
      lat = _ref.lat,
      lon1 = _ref.lon1,
      lat1 = _ref.lat1;
  return lon && lat && lon1 && lat1 ? "".concat(describeLat(lat), ", ").concat(describeLon(lon), " to ").concat(describeLat(lat1), ", ").concat(describeLon(lon1)) : '';
};

var circle = function circle(_ref2) {
  var distance = _ref2.distance,
      unit = _ref2.unit,
      lat = _ref2.lat,
      lon = _ref2.lon;
  return distance && unit && lat && lon ? "".concat(distance, " ").concat(unit, " from ").concat(lat, ", ").concat(lon) : '';
};

var location = function location(_ref3) {
  var lat = _ref3.lat,
      lon = _ref3.lon;
  return lat && lon ? "".concat(lat, ", ").concat(lon) : '';
};

var anyvalue = function anyvalue() {
  return 'is anywhere';
};

var describeNull = function describeNull() {
  return 'is null';
};

var describeNotNull = function describeNotNull() {
  return 'is not null';
};

var filterToStringMap = {
  box: box,
  circle: circle,
  location: location,
  anyvalue: anyvalue,
  "null": describeNull,
  notnull: describeNotNull,
  user_attribute: describeUserAttribute
};
export var describeLocation = function describeLocation(item) {
  return defaultTo(filterToStringMap[item.type], function () {
    return '';
  })(item);
};
