// @looker/visualizations-visx/cjs/Bar/Bar.js
"use strict";

require("core-js/modules/es.array.filter.js");

require("core-js/modules/es.object.get-own-property-descriptors.js");

require("core-js/modules/es.array.index-of.js");

require("core-js/modules/es.weak-map.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Bar = void 0;

require("core-js/modules/es.array.map.js");

require("core-js/modules/es.array.iterator.js");

var _react = _interopRequireWildcard(require("react"));

var _xychart = require("@visx/xychart");

var _text = require("@visx/text");

var _visualizationsAdapters = require("@looker/visualizations-adapters");

var _XYLegend = require("../XYLegend");

var _isArray = _interopRequireDefault(require("lodash/isArray"));

var _get = _interopRequireDefault(require("lodash/get"));

var _compact = _interopRequireDefault(require("lodash/compact"));

var _utils = require("../utils");

var _XYTooltip = require("../XYTooltip");

var _Grid = require("../Grid");

const _excluded = ["formattedValue"],
      _excluded2 = ["formattedValue"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const Bar = ({
  data,
  config,
  height = _visualizationsAdapters.DEFAULT_HEIGHT,
  width,
  fields
}) => {
  var _config$y_axis, _config$x_axis;

  const positioning = config.positioning,
        seriesList = config.series,
        legend = config.legend;
  const formattedData = (0, _utils.concatDimensions)(data, fields);
  const chartTheme = (0, _utils.useChartTheme)(seriesList);
  const visxTheme = (0, _react.useContext)(_xychart.ThemeContext);
  const yAxisLabels = formattedData.map(datum => (0, _utils.formatDateLabel)({
    dateString: datum.dimension,
    fields
  }));
  const yAxisLongestLabel = (0, _visualizationsAdapters.pickLongestLabel)(yAxisLabels);

  const _useMeasuredText = (0, _visualizationsAdapters.useMeasuredText)(yAxisLongestLabel, {
    fontFamily: visxTheme.axisStyles.x.bottom.tickLabel.fontFamily || 'Roboto',
    fontSize: visxTheme.axisStyles.x.bottom.tickLabel.fontSize || '1rem'
  }),
        longestLabelWidth = _useMeasuredText.width;

  if (!(0, _utils.isValidChartData)(data, fields)) {
    return null;
  }

  const renderedBars = (0, _compact.default)(fields.measures.map((measure, i) => {
    const series = (0, _isArray.default)(seriesList) ? (0, _get.default)(config, ['series', i]) : (0, _get.default)(config, ['series', measure.name]);
    if (!series.visible) return undefined;
    return _react.default.createElement(_xychart.BarSeries, {
      key: i,
      dataKey: measure.name,
      data: formattedData,
      xAccessor: d => (0, _utils.getY)(d, i),
      yAccessor: d => (0, _utils.getX)(d)
    });
  }));
  const domain = positioning === 'percent' ? [0, 1] : (0, _utils.getYAxisRange)({
    config,
    data: formattedData,
    fields
  });

  const xScale = _objectSpread({
    type: 'linear'
  }, domain && {
    domain,
    zero: false
  });

  const xAxisConfig = config === null || config === void 0 ? void 0 : (_config$y_axis = config.y_axis) === null || _config$y_axis === void 0 ? void 0 : _config$y_axis[0];

  const xAxis = _react.default.createElement(_xychart.Axis, {
    hideTicks: !(xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values),
    label: (xAxisConfig === null || xAxisConfig === void 0 ? void 0 : xAxisConfig.label) || '',
    labelOffset: xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values ? undefined : 0,
    labelProps: _objectSpread({}, visxTheme.axisStyles.x.bottom.axisLabel),
    orientation: "bottom",
    tickComponent: _ref => {
      let formattedValue = _ref.formattedValue,
          tickProps = _objectWithoutProperties(_ref, _excluded);

      return xAxisConfig !== null && xAxisConfig !== void 0 && xAxisConfig.values ? _react.default.createElement(_text.Text, tickProps, formattedValue) : null;
    }
  });

  const yAxisConfig = config === null || config === void 0 ? void 0 : (_config$x_axis = config.x_axis) === null || _config$x_axis === void 0 ? void 0 : _config$x_axis[0];

  const yAxis = _react.default.createElement(_xychart.Axis, {
    hideTicks: !(yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values),
    label: (yAxisConfig === null || yAxisConfig === void 0 ? void 0 : yAxisConfig.label) || '',
    labelOffset: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? undefined : 0,
    labelProps: _objectSpread(_objectSpread({}, visxTheme.axisStyles.y.left.axisLabel), {}, {
      dx: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? -longestLabelWidth - 10 : -10
    }),
    orientation: "left",
    tickComponent: _ref2 => {
      let formattedValue = _ref2.formattedValue,
          tickProps = _objectWithoutProperties(_ref2, _excluded2);

      return yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? _react.default.createElement(_text.Text, tickProps, (0, _utils.formatDateLabel)({
        dateString: formattedValue || '',
        fields
      })) : null;
    }
  });

  const chartMargin = {
    right: 0,
    top: 0,
    bottom: 50,
    left: yAxisConfig !== null && yAxisConfig !== void 0 && yAxisConfig.values ? longestLabelWidth + 50 : 50
  };
  return _react.default.createElement(_xychart.DataProvider, {
    xScale: xScale,
    yScale: {
      type: 'band'
    },
    theme: chartTheme
  }, _react.default.createElement(_visualizationsAdapters.VisWrapper, {
    legend: legend
  }, _react.default.createElement(_xychart.XYChart, {
    margin: chartMargin,
    width: width,
    height: height
  }, xAxis, yAxis, _react.default.createElement(_Grid.Grid, {
    config: config
  }), _react.default.createElement(_XYTooltip.XYTooltip, {
    config: config,
    data: formattedData,
    fields: fields,
    showDatumGlyph: false,
    snapToDatum: false
  }), positioning === 'stacked' || positioning === 'percent' ? _react.default.createElement(_xychart.BarStack, {
    offset: positioning === 'percent' ? 'expand' : 'none'
  }, renderedBars) : _react.default.createElement(_xychart.BarGroup, null, renderedBars)), _react.default.createElement(_XYLegend.XYLegend, {
    config: config,
    fields: fields
  })));
};

exports.Bar = Bar;
