// @looker/visualizations-adapters/cjs/__mocks__/mockQueryContextValues.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockContextWithNull = exports.mockQueryContextValues = void 0;

var _mockConfig = require("./mockConfig");

var _mockFields = require("./mockFields");

const mockQueryContextValues = {
  config: _mockConfig.mockBarConfig,
  data: [{
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 1088,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2515,
    'orders.average_total_amount_of_order_usd': 1069,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }],
  fields: _mockFields.mockFields,
  ok: true,
  loading: false
};
exports.mockQueryContextValues = mockQueryContextValues;
const mockContextWithNull = {
  config: _mockConfig.mockBarConfig,
  data: [{
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 1088,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2087,
    'orders.average_total_amount_of_order_usd': 88,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': null,
    'orders.average_total_amount_of_order_usd': null,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 3087,
    'orders.average_total_amount_of_order_usd': 88,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }, {
    'orders.count': 2515,
    'orders.average_total_amount_of_order_usd': 1069,
    'orders.created_date': '2019-12-22',
    'users.state': 'California'
  }],
  fields: _mockFields.mockFields,
  ok: true,
  loading: false
};
exports.mockContextWithNull = mockContextWithNull;
