// @looker/visualizations-visx/cjs/XYLegend/XYLegend.js
"use strict";

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.weak-map.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.XYLegend = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = require("styled-components");

var _xychart = require("@visx/xychart");

var _legend = require("@visx/legend");

var _partial = _interopRequireDefault(require("lodash/partial"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const DEFAULT_LEGEND_WIDTH = 200;

const XYLegend = ({
  config,
  fields
}) => {
  const _useContext = (0, _react.useContext)(_xychart.DataContext),
        colorScale = _useContext.colorScale,
        visxTheme = _useContext.theme,
        margin = _useContext.margin;

  const _useContext2 = (0, _react.useContext)(_styledComponents.ThemeContext),
        small = _useContext2.space.small;

  const legend = config.legend;

  if (!legend) {
    return _react.default.createElement(_react.default.Fragment, null);
  }

  const yAxisSpacer = margin === null || margin === void 0 ? void 0 : margin.left;

  const _ref = legend.position === 'left' || legend.position === 'right' ? {
    direction: 'column',
    width: DEFAULT_LEGEND_WIDTH
  } : {
    direction: 'row',
    width: `calc(100% - ${yAxisSpacer})`
  },
        direction = _ref.direction,
        width = _ref.width;

  return _react.default.createElement(_legend.LegendOrdinal, {
    direction: direction,
    itemMargin: `0 ${small} ${small} ${small}`,
    labelFormat: (0, _partial.default)(_utils.seriesLabelFormatter, fields, config),
    style: {
      color: visxTheme === null || visxTheme === void 0 ? void 0 : visxTheme.svgLabelBig.fill,
      display: 'flex',
      justifyContent: 'center',
      marginLeft: yAxisSpacer,
      width
    },
    scale: colorScale,
    shape: "line"
  });
};

exports.XYLegend = XYLegend;
